programs/vote/src/vote_state/mod.rs

```diff
@@ -870,9 +891,33 @@ impl VoteState {
         // `new_vote_state` passed all the checks, finalize the change by rewriting
         // our state.
         if self.root_slot != new_root {
-            // TODO to think about: Note, people may be incentivized to set more
-            // roots to get more credits, but I think they can already do this...
-            self.increment_credits(epoch);
+            // Award vote credits based on the number of slots that were voted on and have reached finality
+            if feature_set
+                .map(|feature_set| feature_set.is_active(&feature_set::timely_vote_credits::id()))
+                .unwrap_or(false)
+            {
+                if let Some(in_slot) = in_slot {
+                    // Credits are awarded based on the "timeliness" of the vote transaction which landed them.  For
+                    // each finalized slot, there was one voted-on slot in the new vote state that was responsible for
+                    // finalizing it.  Each of those votes is awarded credits based on distance from the slot in which
+                    // the vote transaction has landed.
+                    let mut credits = 0;
+
+                    for lockout in new_state.iter().rev() {
+                        credits += Self::compute_vote_slot_credits(in_slot, lockout.slot);
```

Comment:
Right, the mechanism for counting credits is not "perfect" in that it doesn't actually award credits based on the timeliness of every vote exactly. It awards it based on timeliness of the vote that resulted in a dequeue of a prior vote.  I think this still gives incentive for voting with minimal latency, since extra latency can only make credit earnings worse.  But it is also the case that "unlucky" sequences can result in less credit earnings than would otherwise be expected, such as:

A validator that has great latency for some votes but then there is a block gap and those votes expire, and then the validator has "worse" latency for later votes.  They don't get credit for the lesser latency that they achieved for the expired votes, instead they get the worse latency for the later votes.

In addition, when there are vote gaps due to expiry, the votes that survived before the expired votes necessarily end up with fewer credits.  This does add some advantage to a voting strategy that works to ensure that votes are expired as rarely as possible.  This includes strategies that have much more aggressive "failed switch threshold" behaviors, i.e. that commit more heavily to a fork by voting more frequently.  

An alternative that I considered was tracking the voted-in slot along with every vote, that way when a vote is dequed credits can be awarded based on the latency of the vote that originally cast it, which I think is the most accurate accounting possible.  However, I believe that this would require changing the tower's persisted data structure and I thought that was likely to be too risky a change.